In [1]:
#product_launch.xlsx
# The goal is to create a multiple linear regression model
# that reliably generates new data.
import pandas as pd
import micropip
await micropip.install('pandas')
await micropip.install('openpyxl')

# excel file containing the product launch data
excel_file = pd.ExcelFile("practice_RMSE.xlsx", engine='openpyxl')
# read the file into a pandas data frame
data = pd.read_excel(excel_file)

# Build a multiple linear regression model using k-fold cross-validation and 
# use RMSF as the assessment measure.
# Import the cross_val_score function for performing cross validation
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.tree import DecisionTreeRegressor

# select the features and target variable
features = ['Pianos_Tuned (#)', 'Advertising_Expenses', 'Maintenance_Costs']
target = 'Monthly_Revenue'
# prepare the future matrix (x) and target vector (y)
x = data[features]
y = data[target]

# initialize the model
model = LinearRegression()
# set a random seed for reproducibility
# shuffling process arranges the data in the same way every time
np.random.seed(42)
# preform 5-fold cross-validation
mse_scores = cross_val_score(model, x, y, cv=5, scoring='neg_mean_squared_error')
# calculate the RMSE across all folds
rmse_scores = (-mse_scores) ** 0.5
average_rmse = rmse_scores.mean()
# print the average RMSE
print(f"Average Root Mean Squared Error from 5-fold Cross-Validation: {average_rmse:.0f}")

Average Root Mean Squared Error from 5-fold Cross-Validation: 108


In [2]:
# Initialize the model
tree_model = DecisionTreeRegressor()

# Cross-validation: 5 folds
tree_mse_scores =  cross_val_score(tree_model, x, y, cv=5, scoring='neg_mean_squared_error')

# Convert MSE to RMSE
tree_rmse_scores = (-tree_mse_scores) ** 0.5

# compute average RMSE for both models
tree_avg_rmse = tree_rmse_scores.mean()

# Print results
print(f'Average RMSE from Decision Tree: {tree_avg_rmse:.0f}')

Average RMSE from Decision Tree: 332


After evaluating both the multiple linear regression model and the decision tree model using 5-fold crossvalidation, it is observed that the multiple linear regression model has an average RMSE of 108 points, while the
decision tree model has an average RMSE of 332 points.
This indicates that the multiple linear regression model provides more accurate predictions with lower error
compared to the decision tree model for this particular data set. Therefore, you should recommend using the
multiple linear regression model for predicting the target variable, as it demonstrates better performance in this
scenario.